### DISCUSSION TITLE
computational efficiency

### DISCUSSION THREAD

**Discussion by ddbuscombe** [Doodleverse/segmentation_gym#45] [Ideas]
a memory bottleneck in multiclass problems is caused by the use of one-hot encoded label stacks

binary problems can use larger TARGET_SIZE and / or BATCH_SIZE than multiclass problems, for this reason

One possible workaround is to train separate binary models for each class in a multiclass problem

**Comment by ebgoldstein** [comment]
another idea is to use 'label encoding' (not one-hot) for labels and use `SparseCategoricalCrossentropy` as loss.

So last layer is still: `Batch * X_dim * Y_dim * N_Classes` but the labels can be sent through as `Batch * X_dim * Y_dim * 1`

**Comment by ddbuscombe** [comment]
- [ ] merge branch containing distributed training 
- [ ] implement sparse loss (requires flattened 2d labels, not one-hot encoded)
- [ ] implement parallel processing for `seg_images_in_folder`

**Comment by ddbuscombe** [comment]
when we're ready to work on that, this should be converted to an issue

**Comment by ddbuscombe** [comment]
I tried implementing some ideas listed [here](https://stackoverflow.com/questions/60905801/parallelizing-model-predictions-in-keras-using-multiprocessing-for-python) but no joy

I also tried to put all the model building commands inside the `do_seg` function ... this would make sense if set_gpu= False, however I cant use my CPU with tensorflow on my main development machine, so I cant test it
